rosidl_runtime_rs/src/sequence.rs

```diff
@@ -0,0 +1,618 @@
+/// ```
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct BoundedSequence<T: SequenceAlloc, const N: usize> {
+    inner: Sequence<T>,
+}
+
+/// Error type for [`BoundedSequence::try_new()`].
+#[derive(Debug)]
+pub struct SequenceExceedsBoundsError {
+    len: usize,
+    upper_bound: usize,
+}
+
+/// A by-value iterator created by [`Sequence::into_iter()`] and [`BoundedSequence::into_iter()`].
+pub struct SequenceIterator<T: SequenceAlloc> {
+    seq: Sequence<T>,
+    idx: usize,
+}
+
+// ========================= impl for Sequence =========================
+
+impl<T: SequenceAlloc> Clone for Sequence<T> {
+    fn clone(&self) -> Self {
+        let mut seq = Self::default();
+        if T::sequence_copy(self, &mut seq) {
+            seq
+        } else {
+            panic!("Cloning Sequence failed")
+        }
+    }
+}
+
+impl<T: Debug + SequenceAlloc> Debug for Sequence<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        self.as_slice().fmt(f)
+    }
+}
+
+impl<T: SequenceAlloc> Default for Sequence<T> {
+    fn default() -> Self {
+        Self {
+            data: std::ptr::null_mut(),
+            size: 0,
+            capacity: 0,
+        }
+    }
+}
+
+impl<T: SequenceAlloc> Deref for Sequence<T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        unsafe { std::slice::from_raw_parts(self.data, self.size) }
+    }
+}
+
+impl<T: SequenceAlloc> DerefMut for Sequence<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { std::slice::from_raw_parts_mut(self.data, self.size) }
+    }
+}
+
+impl<T: SequenceAlloc> Drop for Sequence<T> {
+    fn drop(&mut self) {
+        T::sequence_fini(self)
+    }
+}
+
+impl<T: SequenceAlloc + Eq> Eq for Sequence<T> {}
+
+impl<T: SequenceAlloc> Extend<T> for Sequence<T> {
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item=T>,
+    {
+        let mut it = iter.into_iter();
+        // The index in the sequence where the next element will be stored
+        let mut cur_idx = self.size;
+        // Convenience closure for resizing self
+        let resize = |seq: &mut Self, new_size: usize| {
+            let old_seq = std::mem::replace(seq, Sequence::new(new_size));
+            for (i, elem) in old_seq.into_iter().enumerate().take(new_size) {
+                seq[i] = elem;
+            }
+        };
+        // First, when there is a size hint > 0 (lower bound), make room for
+        // that many elements.
+        let num_remaining = it.size_hint().0;
+        if num_remaining > 0 {
+            let new_size = self.size.saturating_add(num_remaining);
+            resize(self, new_size);
+        }
+        while let Some(item) = it.next() {
+            // If there is no more capacity for the next element, resize to the
+            // next power of two.
+            //
+            // A pedantic implementation would check for usize overflow here, but
```

Comment:
Should we be pedantic, though? If `ros2_rust` is being used on safety-critical systems, they may want this sort of check. Or is there a better value to test against? What would the performance impact be?

Reply:
Yeah, good question. I'm not worried about performance impact, but about code complexity/overengineering.

I think ROS 2 messages will never reach this limit – certainly not on 64-bit machines, and I don't know if anyone is building ROS 2 for 32-bit systems. There is also a 1 GB message size limit, so even if we have a 32-bit system, there won't be any sequences with more than `usize/2` elements. Besides, there are probably plenty of unchecked additions in rcl and rmw that would cause trouble too.

So I propose to not be pedantic about usize overflow, but to just leave comments like this one when such an issue is encountered. If anyone ever really wants to use `ros2_rust` in a safety-critical context, they'll need to go through the code and refactor a whole lot of stuff anyway, and then at least there's a comment to point to potential issues.